Reorganize Maven modules to isolate functionality better

wetneb opened the pull request:
This is the first PR in an attempt to break my work on 4.0 into reviewable chunks.

This PR does not contain functional changes: it just rearranges our Java classes into more Maven modules.
So far, most of our code was in the `main` Maven module. Breaking this down into more modules has a couple of benefits:
* making it possible for developers to work on a smaller part of the code-base without interfering with other work
* controlling the inter-dependencies between features of the tool, to avoid accidental dependencies for instance introduced by unit tests which rely on unrelated functionalities to set up a test scenario
* making it possible for dependents (such as extensions) to declare dependencies on the parts of our code they really depend on
* exposing test utilities in a standalone module, so that extensions can rely on it without having to copy over test utilities from our code base
* potentially, enabling concurrent compilation and testing

wetneb committed:
Cleanup Maven dependencies

wetneb committed:
Remove duplication in Wikidata extenison

wetneb committed:
Fix Maven packaging phase for new modules, fix test

wetneb committed:
Remove unused ChangeSequence

wetneb committed:
Refactor powermock dependency definition

wetneb committed:
Reorganize 'main' module to follow Maven conventions

wetneb committed:
Remove dependency from or-util to or-model

tfmorris reviewed (commented):
I've done a cursory review of this chunk of commits and it confirmed my worst fears about the reviewability of giant PRs. In a commit labeled "Extract GREL module" there is buried a change removing multi-character delimiter support from SmartSplit. I stumbled across this mostly by accident and fear that other similar things, whether intentional changes or unintentional mistakes from merges, etc, are going to be impossible to find buried in even larger PRs

wetneb commented on or-model/src/main/java/com/google/refine/RefineModel.java:
```diff
@@ -0,0 +1,87 @@
+    }
+    
+    static {
+        registerClassMapping("com.metaweb.*", "com.google.*");
+        registerClassMapping("com.google.gridworks.*", "com.google.refine.*");
```
I agree, but that should be a separate change, isn't it?

wetneb commented on main/webapp/modules/core/MOD-INF/controller.js:
```diff
@@ -314,12 +314,28 @@ function registerImporting() {
 function init() {
   // Packages.java.lang.System.err.println("Initializing by script " + module);
 
+  registerLanguages();
+  registerFacets();
```
Those are not new extension points, it consists in moving where those components are registered (before, they were registered statically in the java code). Extensions were already able to extend those components by registering their components in a similar way. I just think it's cleaner if this is done via the `controller.js`, as it serves that purpose. Of course, it's still not very satisfactory: we would be better off with a declarative version of those component registrations, but that's a more involved change.

wetneb commented on or-model/src/main/java/com/google/refine/browsing/facets/FacetConfigResolver.java:
```diff
@@ -46,16 +54,25 @@ public Id getMechanism() {
     public JavaType typeFromId(DatabindContext context, String id) throws IOException {
-        return factory.constructSimpleType(ReconConfig.getClassFromMode(id), new JavaType[0]);
+    	// backwards compatibility
+    	if (id.indexOf('/') == -1) {
+    		id = "core/"+id;
+    	}
```
It makes sense to add a test covering this case.

wetneb commented on or-model/src/main/java/com/google/refine/expr/MetaParser.java:
```diff
@@ -70,13 +69,6 @@ static public class LanguageInfo {
 //    final static private Var CLOJURE_EVAL = RT.var("clojure.core", "eval");
     
     static {
-        registerLanguageParser("grel", "General Refine Expression Language (GREL)", new LanguageSpecificParser() {
```
Because I moved GREL to its own module and I did not think it was worth doing that for Clojure as it's a much smaller implementation. Would you like to have Clojure in its own Maven module as well?

wetneb commented on or-model/src/test/java/com/google/refine/history/HistoryEntryTests.java:
```diff
@@ -34,27 +34,21 @@
-            + "   \"baseColumnName\":\"country\","
-            + "   \"expression\":\"grel:\\\"https://www.wikidata.org/wiki/\\\"+cell.recon.match.id\","
-            + "   \"onError\":\"set-to-blank\"}"
+            + "\"operation\":{\"op\":\"core/my-operation\","
```
Because the previous version of this test made references to a specific operation and GREL expression. Those are features which are now defined in other Maven modules, which are not accessible in this scope. `HistoryEntry` is part of the core model of OpenRefine (`or-model`), on which GREL (`or-grel`) and workflow elements (`or-workflow`) depend. Therefore it is necessary to introduce a stub operation in this case. From a unit testing perspective, it also makes more sense to me since in this test we are not meant to assert anything about how a specific operation is serialized, but rather that the serialization of `HistoryEntry` relies on the serialization of the contained operation.

wetneb commented on pom.xml:
```diff
@@ -12,6 +12,8 @@
 
   <modules>
+        <module>or-model</module>
+        <module>or-testing</module>
```
You could merge it with `or-model`, but I think it would be a bit sad to do that, since `or-model` would then have to depend on TestNG / Mockito and the like. The goal of this artifact is to gather all testing utilities that modules in OpenRefine or in extensions can rely on, without having to copy and paste those into their own module.

wetneb commented on main/tests/server/src/com/google/refine/browsing/facets/ListFacetTests.java:
```diff
@@ -99,9 +100,11 @@ public void serializeListFacetConfig() throws JsonParseException, JsonMappingExc
-        Project project = createCSVProject("Column A\n" +
-                "foo\n" +
-                "bar\n");
+        Project project = createProject(new String[]
```
The `createCSVProject` method is one I introduced some years ago as a convenience method to quickly create a project in the context of a test. This had a couple of problems:
* internally, it relies on our CSV importer, which is not very clean: this is a rather complex component, which would get run by a lot of different tests, even though those corresponding tests have nothing to do with it (they test something unrelated).
* specifying a project in CSV form is not very readable in a Java source file, it forces the human reader to parse the CSV visually to understand the structure of the project being created
* creating a project from a CSV is not very flexible: for instance, it is complicated to specify a project with a mixture of empty strings and null values, or with parsed numbers and numbers as strings

So that's why I introduced this new method, which lets testers create projects without going through the CSV importer. The reason why this change is done in the same go as the Maven structure change is that this structure change prevents us from using the CSV importer in testing utilities, because the testing utilities module does not depend on the workflow elements module.

wetneb commented on main/tests/server/src/com/google/refine/model/recon/StandardReconConfigTests.java:
```diff
@@ -61,8 +62,8 @@ public class StandardReconConfigTests extends RefineTest {
     public void registerOperation() {
-        OperationRegistry.registerOperation(getCoreModule().getName(), "recon", ReconOperation.class);
-        ReconConfig.registerReconConfig(getCoreModule().getName(), "standard-service", StandardReconConfig.class);
+        OperationRegistry.registerOperation("core", "recon", ReconOperation.class);
+        ReconConfig.registerReconConfig("core", "standard-service", StandardReconConfig.class);
```
I am keen to migrate to something else, but I think it's a bigger change indeed (for instance because it requires changes to Butterfly itself): https://forum.openrefine.org/t/improving-the-ux-of-extension-install-and-butterfly/52

wetneb commented on main/src/com/google/refine/clustering/knn/DistanceFactory.java:
```diff
@@ -54,19 +45,7 @@ public class DistanceFactory {
 	private static List<String> _distanceNames = new LinkedList<>();
     
     static {
-        put("levenshtein", new VicinoDistance(new LevenshteinDistance()));
-        put("ppm", new VicinoDistance(new PPMDistance()));
```
Adding initialization to the test cases does not feel that unusual to me: that's something we have been doing for operations for a long time. I think it's a good practice in that it makes sure the test declares all the components it depends on and therefore avoids accidental dependencies which should be avoided (typically, in the case you pointed out above, `HistoryEntryTests` depending on GREL).

wetneb commented on main/webapp/modules/core/MOD-INF/controller.js:
```diff
@@ -328,6 +328,30 @@ function registerFacets() {
   FCR.registerFacetConfig("core", "scatterplot", Packages.com.google.refine.browsing.facets.ScatterplotFacet.ScatterplotFacetConfig);
 }
 
+function registerDistances() {
+   var DF = Packages.com.google.refine.clustering.knn.DistanceFactory;
+   var VicinoDistance = Packages.com.google.refine.clustering.knn.VicinoDistance;
+   DF.put("levenshtein", new VicinoDistance(new Packages.edu.mit.simile.vicino.distances.LevenshteinDistance()));
+   DF.put("ppm", new VicinoDistance(new Packages.edu.mit.simile.vicino.distances.PPMDistance()));
+        
+   // Distances not activated as they are not very useful:
+   // See https://github.com/OpenRefine/OpenRefine/pull/1906
+   /*
+   DF.put("jaccard", new VicinoDistance(new JaccardDistance()));
+   DF.put("jaro", new VicinoDistance(new JaroDistance()));
+   DF.put("jaro-winkler", new VicinoDistance(new JaroWinklerDistance()));
+   DF.put("jaro-winkler-tfidf", new VicinoDistance(new JaroWinklerTFIDFDistance()));
+   DF.put("gzip", new VicinoDistance(new GZipDistance()));
+   DF.put("bzip2", new VicinoDistance(new BZip2Distance()));
+   */
+}
+
+function registerClusterers() {
+   var CCF = Packages.com.google.refine.clustering.ClustererConfigFactory;
+   CCF.register("knn", Packages.com.google.refine.clustering.knn.kNNClusterer);
+   // Binning clusterer: already registered by default.
+}
```
I am very supportive to move to a different extension mechanism, and I would be keen to do this before the new architecture is included in a stable release.

wetneb commented on or-testing/src/main/java/com/google/refine/RefineTest.java:
```diff
@@ -131,63 +115,75 @@ protected Project createProjectWithColumns(String projectName, String... columnN
-        projects.add(project);
-        importingJobs.add(job);
-        return project;
+    protected Project createProject(String projectName, String[] columns, Serializable[] rows) {
```
See my comment there: https://github.com/wetneb/OpenRefine/pull/334#discussion_r1356165051

wetneb commented on or-model/src/main/java/com/google/refine/expr/WrappedRow.java:
```diff
@@ -58,13 +58,9 @@ public Object getField(String name, Properties bindings) {
-        } else if ("columnNames".equals(name)) {
-            Project project = (Project) bindings.get("project");
-            
+        } else if ("columnNames".equals(name)) {           
```
That's something I'll look into more closely. There are bigger changes about the cross function later on, also because we did a refactoring of it in the master branch since then.

wetneb commented on or-grel/src/test/java/com/google/refine/expr/functions/strings/SmartSplitTests.java:
```diff
@@ -96,26 +79,8 @@ public void testSmartSplitCharSepSpace() {
-        } else {
-            return function.call(bindings, args);
-        }
+        Object actual = invoke("smartSplit", testString, "#@$");
+        assertTrue(actual instanceof EvalError);
```
For me that was a conscious and intended change: from my perspective multi-char support for the smartSplit function is mostly a side-effect of us wanting to support multi-char CSV support at the import stage, but I don't think it is critical in this GREL context. It's fair to disagree, and I'd be happy to restore this behaviour via commons-csv for instance. For sure, I should have done a better job at announcing this change, but honestly I think in the context of such large-scale architecture changes we need to have a proportionate approach and accept small differences in functionality - otherwise we are just making it impossible for ourselves to do significant renovation work on the code base.

wetneb replied:
I have rebased my changes on top of the master branch, after your univocity migration, so that's resolved.

wetneb commented on or-grel/src/main/java/com/google/refine/expr/functions/strings/SmartSplit.java:
```diff
@@ -71,15 +73,12 @@ public Object call(Properties bindings, Object[] args) {
+            		String sep = args[1].toString();
+                	parser = buildParser(sep.charAt(0));
+            	} else {
+            		return new EvalError(ControlFunctionRegistry.getFunctionName(this) + " only supports single-character separators");
```
See response above.

wetneb committed:
Separation of the core datamodel into a standalone Maven package.

This is a breaking change since it changes extension points to avoid references
to Butterfly in the registration of operations and recon configs.

wetneb committed:
Extract GREL Maven module

wetneb committed:
Fix compilation error in benchmark module
